### PR TITLE
highlighting graphical expressions and their visual elements

### PR DESCRIPTION
**Pull request by luminaxster**

### CODE REVIEW

File: src/htmlViewer/html-viewer.js [hunk 1]
```diff
@@ -92,15 +92,15 @@ export class HtmlViewer {
...
-    //       $(referenceTimeline[index].reference).css("background-color", "#ffe6e6");
-    //     }
-    //   }
-    //
-    // });
+    ea.subscribe("graphicalTraceChanged", payload => {
+      let referenceTimeline = payload;
+      for (let index in referenceTimeline) {
+        if (referenceTimeline[index].isGraphical) {
+          $(referenceTimeline[index].reference).css("background-color", "#ffe6e6");
```
Comment (luminaxster) [inline src/htmlViewer/html-viewer.js]: Here we look all graphical expressions and highlight their visual elements

File: src/visualAnalysis/graphical-analyzer.js [hunk 2]
```diff
@@ -3,30 +3,44 @@
...
+    this.eventAggregator.subscribe("expressionAfterCursorChanged", match =>{
+      // you should only proceed if you have graphic timeline
+      if(!this.graphicalTimeline){
+        return;
+      }
+      // and you have an expression match
+      if(!match){
+        return;
+      }
+      // here you get the info for expressions after cursor changes
```
Comment (luminaxster) [inline src/visualAnalysis/graphical-analyzer.js]: at this point you can always get the expression the developer is typing after the trace is valid

**Comment by Varun-Kulkarni** [comment]
Different colored highlights for different graphical references' code are not showing, however if I use the `default-marker` it updates for all lines.  I feel like there is some issue in the ace marker registry or maybe the way I declared each marker but I'm not sure

**Comment by Varun-Kulkarni** [comment]
Figured it out.. "ace_marker-layer" not "ace-marker-layer"

**Comment by luminaxster** [comment]
Yep. Ace violates HTML naming conventions =[

File: src/htmlViewer/html-viewer.js [hunk 3]
```diff
@@ -92,15 +94,32 @@ export class HtmlViewer {
...
+      //todo create colors based on ace editor highlights
+      /**Create Colors for each unique reference**/
+      // let num = (1 / (uniqueReferences.length));
+      // let op = num;
+      // for(let index in uniqueReferences){
+      //   let obj = uniqueReferences[index];
+      //   op += num;
+      // }
+    });
+
```
Comment (luminaxster) [inline src/htmlViewer/html-viewer.js]: subscribe to highlightVisualElement, receives the DOM reference and the highlight color, if empty restore original background 
reference.css("background-color", color);

File: src/visualAnalysis/graphical-analyzer.js [hunk 4]
```diff
@@ -3,30 +3,146 @@
...
+      //   }
+      //   let graphicalObjString = createToken(superRef[0]);
+      //   console.log("check", hoveredDOMString === graphicalObjString, "Hovered String: " + hoveredDOMString, "Graphical String: " + graphicalObjString);
+      //
+      //   // console.log("names: " + superRef, createToken(superRef[0]), name, superRef === name);
+      // }
+      // //Todo: Something went wrong in the code above because the references are really off plz check
+
+    });
+
```
Comment (luminaxster) [inline src/visualAnalysis/graphical-analyzer.js]: subscribe to "expressionHovered" and look for the range in the referenceTimeline. Then publish "highlightVisualElement" with the DOM reference and the background color

**Comment by luminaxster** [comment]
I provide hints for highlighting the visual elements as well

**Comment by Varun-Kulkarni** [comment]
So currently I've tried to use the code suggested by StackOverflow which uses a style variable and  `document.createElement()`.  I add this style to the array of markers to be used, however when I call aceUtils to make the marker manager and such, it doesn't seem to be able to read the class.  Not sure why

**Pull request closed without merging**